### ISSUE TITLE
Upgrade Istanbul version

### ISSUE THREAD

**Issue by simondel** [stryker-mutator/stryker#524]
The current version of Istanbul is quite old and no longer maintained. The repo has been moved from:
https://github.com/gotwarlost/istanbul
to:
https://github.com/istanbuljs/istanbuljs

It has also been split up into multiple smaller packages. Perhaps we can also shrink down what we use from Istanbul.

This also fixes #521 and may even help us to add coverage analysis for TypeScript because nyc can already do this and nyc uses the new istanbul packages.

**Comment by simondel** [comment]
We can use the package[istanbul-lib-instrument](https://github.com/istanbuljs/istanbuljs/tree/master/packages/istanbul-lib-instrument) which exposes a function `createInstrumenter` to create an instrumenter. The instrumented code differs from the previous istanbul version in a couple of ways:
```js
var cov_wsa02hu9w = function() {
    var path = "file.js",
        hash = "843b8a51a856c163118cd8cb7c853e4a5b66e222",
        global = new Function('return this')(),
        gcv = "__strykerCoverageCurrentTest__",
        coverageData = {
            path: "file.js",
            statementMap: {
                "0": {
                    start: {
                        line: 1,
                        column: 0
                    },
                    end: {
                        line: 1,
                        column: 5
                    }
                }
            },
            fnMap: {},
            branchMap: {},
            s: {
                "0": 0
            },
            f: {},
            b: {},
            _coverageSchema: "332fd63041d2c1bcb487cc26dd0d5f7d97098a6c"
        },
        coverage = global[gcv] || (global[gcv] = {});
    if (coverage[path] && coverage[path].hash === hash) {
        return coverage[path];
    }
    coverageData.hash = hash;
    return coverage[path] = coverageData;
}();
cov_wsa02hu9w.s[0]++;
a + b;
```

1. The coveragedata object contains property names that are not surrounded by strings such as `"path"` in the previous istanbul version. This causes our regex to fail
2. The coverage object hsa been renamed to `cov_{funtioname}` instead of `__cov_`
3. The statement maps seems to start at 0 now instead of 1 and also called using a number instead of a string.

**Comment by nicojs** [comment]
Thanks for this issue! This has been on my radar for quit some time. We should be careful here. The mutation score should not change after updating (mutation score calculated on the same code).

**Comment by simondel** [comment]
Yeah, we should indeed be careful. It's a risky upgrade. The covered/non-covered ratio should also stay the same.

**Comment by nicojs** [comment]
I've took some time to try it out and have the following findings:

We don't need a regex anymore to get at the statement map. Instead we can use `instrumenter.lastFileCoverage().statementMap`.

I don't believe the coverage object name is changed. We can simply supply the coverage object name: `createInstrumenter({ coverageVariable: 'ourGlobalCoverageObjectName })`.

The statement maps still start at 1, so we still need to correct it: 

```ts
 Object.keys(statementMap).forEach(key => {
      // Lines from istanbul are one-based, lines in Stryker are 0-based
      statementMap[key].end.line--;
      statementMap[key].start.line--;
    });
```

There is an other challenge. A function block is not seen as a statement anymore, which in turn means we have to match a block mutant with every test result and a warning is logged: 

```
[2017-12-12 09:27:11.586] [WARN] MutantTestMatcher - Cannot find statement for mutant [object Object] in statement map for file. Assuming that all tests cover this mutant. This might have a big impact on the performance.
```

In order to fix this, we might need to resort to using the function map in order to match the block statement.

**Comment by nicojs** [comment]
Work for this issue is being done in branch feat/istanbul-lib-instrument

**Comment by simondel** [comment]
Fixed with #550
